cmd/server/main.go

```diff
@@ -541,8 +543,11 @@ func (p *Proxy) runAgentServer(o *ProxyRunOptions, server *server.ProxyServer) e
 	}
 
 	addr := fmt.Sprintf(":%d", o.agentPort)
-	serverOption := grpc.Creds(credentials.NewTLS(tlsConfig))
-	grpcServer := grpc.NewServer(serverOption)
+	serverOptions := []grpc.ServerOption {
+		grpc.Creds(credentials.NewTLS(tlsConfig)),
+		grpc.KeepaliveParams(keepalive.ServerParameters{Time: 30 * time.Second}),
```

Comment:
This is way too short to be the default. Currently the max K8s cluster size is 15K nodes. For such a cluster you would be sending 500 keep-alive packets per second to each server. I would suggest making the default 1 hour (the default TCP keepalive on the server is 2 hours (so 1/2 that is good) and the default server enforced minimum is 5 minutes).  I'm fine with this being a optional command line flag so you can tune this to you needs. I think we are still likely to also want to mark problematic tunnels as unhealthy but having both seems a good idea.

Reply:
I wonder if we also want to set a keep alive on the agent side.

Reply:
Adding an optional command line flag is a good idea - I will figure it out.

As I understand, we don't need a keepalive mechanism on the agent side, because we currently [check health there](https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/3cf58829c4f429fd038b7a5797d112979878060b/pkg/agent/client.go#L436), using [grpc.ClientConn.GetState](https://godoc.org/google.golang.org/grpc#ClientConn.GetState).

Reply:
I *believe* (I could certainly be wrong) that that GetState is mostly tracking if closed has been called. It is not tracking if we have either a network outage such that traffic doesn't route or that the server has gone away without notifying us. These are the conditions which keepalive is meant to help. https://godoc.org/google.golang.org/grpc#WithKeepaliveParams client side keep alive is almost identical to server side keepalive and they should work well together.

Reply:
Just going to add that SSH tunnels have a healthCheckPoll time of 1 minute (https://github.com/kubernetes/kubernetes/blob/master/pkg/ssh/ssh.go#L316), so it's possible that we could go lower than 1 hour for the keepalive time.

We can probably address client side keep alive in a separate PR.

Reply:
I think so, too. ~1 minute keepalive time isn't something bad for performance - remember, that we send keepalive packets only when the channel is inactive for longer than specified time. In a 15K-node cluster, administrators probably will not deploy konnectivity as daemonset, on every node.
I'm currently working on adding a command line option to allow the administrator to tune this parameter.

After I complete this PR, I would like to try to add keepalive to the agent and then create a new PR :)

Reply:
@cheftako, @Jefftree, I've just added a `--keepalive-time` command-line option to the server code and Kubernetes example.